packages/eslint-plugin/src/rules/no-floating-promises.ts

```diff
@@ -296,12 +327,31 @@ export default createRule<Options, MessageId>({
   },
 });
 
+function isPromiseArray(checker: ts.TypeChecker, node: ts.Node): boolean {
+  const type = checker.getTypeAtLocation(node);
+  for (const ty of tsutils
+    .unionTypeParts(type)
+    .map(t => checker.getApparentType(t))) {
+    if (checker.isArrayType(ty)) {
```

Comment:
`isArrayType` covers most cases, but not all. Specifically tuples are trouble. I'd need to add something additional (not sure what exactly) in order to get the following test cases to pass
```
[Promise.reject()] as const; // should error
```
```
declare function cursed(): [Promise<unknown>];
cursed(); // should flag 
```

Reply:
So, yes, but I wasn't sure then how the type arguments/generic types would work (and haven't been able to find any docs).

I'll change it to my best guess, but then I'd appreciate confirmation it's sound.

Reply:
Yeah the APIs for TypeScript are ... not particularly thoroughly documented. We've mostly just stumbled along with what's exposed publicly (or we've felt comfortable writing manual types to make exposed). If it works then I'd say 👍 go for it.